### DISCUSSION TITLE
run in (ram) limitations in berry script?

### DISCUSSION THREAD

**Discussion by ChrisVonFrosta** [arendst/Tasmota#18100] [Q&A]
Hi, i have a 500 lines berry script (13167 bytes).
Since my last update from this script, barry script is no coming up. no message is showing up in the berry console after restart. No cmd is working  in berry console.
i can fix it by comment out one random code line. No matter which code line I comment out, the Berry script starts.

![grafik](https://user-images.githubusercontent.com/16272493/222732521-191ab054-61fa-4925-aa54-e7d27ae180e8.png)

**Comment by Jason2866** [comment]
Use a esp32x device which has PSRAM.

**Reply by ChrisVonFrosta** [reply]
thx! psram means more ram?

**Comment by barbudor** [comment]
In the Tasmota Console (not Berry, the standard console) you can have a look at the STATE message which provides details such as below:
```
{"Time":"2023-03-03T14:27:35+01:00","Uptime":"19T18:55:17","UptimeSec":1709717,"Heap":87,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":2,"Berry":{"HeapUsed":3,"Objects":30},"Wifi":{"AP":1,"SSId":"XXXXXXXX","BSSId":"AA:AA:AA:AA:AA:AA","Channel":1,"Mode":"11n","RSSI":100,"Signal":-3,"LinkCount":1,"Downtime":"0T00:00:06"}}
```
The general `Heap` value is the available RAM (same as in the information page) and the Berry specific information can give you some hints

**Reply by s-hadinger** [reply]
Probably a bootloop due to no more memory. After 4 boots Berry is automatically disabled. This is why the Berry console does not work anymore until next boot

**Reply by ChrisVonFrosta** [reply]
15:22:25.901 RSL: STATE = {"Time":"2023-03-03T15:22:25","Uptime":"0T00:00:07","UptimeSec":7,"Heap":155,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"POWER1":"OFF","POWER2":"OFF","POWER3":"OFF","POWER4":"OFF","POWER5":"OFF","POWER6":"OFF","POWER7":"OFF","POWER8":"OFF","POWER9":"ON","Wifi":{"AP":1,"SSId":"Phradio","BSSId":"B0:F2:08:07:B4:28","Channel":1,"Mode":"11n","RSSI":54,"Signal":-73,"LinkCount":1,"Downtime":"0T00:00:02"}}

**Comment by ChrisVonFrosta** [comment]
## no line commented out (berry not running):

<html><body>


ESP Chip Id | 12317240 (ESP32-D0WD raev.1)
-- | --
Flash Chip Id | 0x16405E (DIO)
Flash Size | 4096 KB
Program Flash Size | 4096 KB
Program Size | 1479 KB
Free Program Space | 1400 KB
Free Memory | 153.1 KB (frag. 29%)
 
Partition safeboot  | 832 KB (used 91%)
Partition app0* | 2880 KB (used 51%)
Partition fs | 320 KB


</body>
</html>

## one line commented out (berry running):

<html><body>


ESP Chip Id | 12317240 (ESP32-D0WD rev.1)
-- | --
Flash Chip Id | 0x16405E (DIO)
Flash Size | 4096 KB
Program Flash Size | 4096 KB
Program Size | 1479 KB
Free Program Space | 1400 KB
Free Memory | 126.9 KB (frag. 13%)
 
Partition safeboot  | 832 KB (used 91%)
Partition app0* | 2880 KB (used 51%)
Partition fs | 320 KB


</body>
</html>

**Reply by ChrisVonFrosta** [reply]
if i implement more funktions as a driver, could it help? Or is only the total code size of significance?

**Reply by barbudor** [reply]
If you mean a Berry driver aka a class derived from Driver ?
It won't make a change, only total quantity of code and variables would impact

However I see the free ram dropping from 155KB to 127KB when Berry code is running doesn't seem to be the problem

So the RAM usage doesn't seems to be the problem
Especially if you get it running by commenting only any 1 line, seems to push in another direction

Any way to refactor and optimize your code ?
Are you using comments that you could remove ?

**Reply by s-hadinger** [reply]
Removing comments will not change memory usage

**Reply by barbudor** [reply]
With 126kB free and only 13% fragmentation, memory doesn't seem to be the problem.

**Reply by ChrisVonFrosta** [reply]
> Any way to refactor and optimize your code ? Are you using comments that you could remove ?
it 
Yes, I am using a number of comments. How can removing some of them help?
The code is certainly optimizable, but I did not consciously write it unoptimized ;)
As for what the refactoring rate means, I am not sure.

**Reply by barbudor** [reply]
[code refactoring](https://en.wikipedia.org/wiki/Code_refactoring)

while @s-hadinger stated that memory usage would not be changed by removing comments, I don't feel it is straight a memory issue, at least not related to code execution, but may be related to source-parsing or loading.
But obviously, as adding a comment make it work, that's probably a useless test

**Reply by sfromis** [reply]
Removing comments is unlikely to help much, if anything. Without knowing anything about what you comment out, what memory usage you have when running, or what symptoms you may be getting, hard to make specific relevant suggestion about what to change.

What you've posted until now does not suggest why you are thinking of it as a RAM limitation issue, as the numbers does not look like that, unless you have code suddenly allocating a lot.

**Reply by ChrisVonFrosta** [reply]
> What you've posted until now does not suggest why you are thinking of it as a RAM limitation issue, as the numbers does not look like that, unless you have code suddenly allocating a lot.

I simply had no idea what it could be, so I put "ram" in parentheses ;)
Anyway, there is no error message and Berry is not running.

**Comment by ChrisVonFrosta** [comment]
For the structure, you can certainly help. I am using this [example](https://tasmota.github.io/docs/Berry-Cookbook/#adding-a-button-to-the-main-menu) as web.be to display buttons in the web UI. When button is triggered, a module function will executed. I would like to keep such functions in another file for clarity. Would I load this additional file as a module? Where would I run the program logic? In web.be? In the module? Or in a higher-level file?

actually stucture:

`web.be`      ->     webbuttons and push trigger (load `web.be and loaded bei `autoexec.be`)

`module`     ->    functions (read sensor values, controll outputs)

`app-logik`    ->  react to sensor values and do stuff even with callback funkction `tasmota.set_timer()`


actually, i have the app-logik in a module.funktion inside the module.
i think that is not the smartest way?!
maybe you could share best practice for something like that.

**Reply by sfromis** [reply]
To me, it does not really look like "something which *should* be refactored". Of course it is always possible to discuss what is a good way of doing things, and what might be done in a better way, but that should not be related to your issue which could very well be due to some bug in Tasmota or Berry tripping up over some unusual situation.

I'd prefer trying to isolate when it happens, like if you can get the failure if you insert print statements at interesting points in the code, to see how far it gets. What you see with resetting does look like what s-hadinger suggested about a boot loop where Tasmota may deable things to at least be able to boot without crashing.

**Comment by ChrisVonFrosta** [comment]
i tried with print("test"). but since berry do not run i had no hope. but...
first i add a print() to the begin of the module.be -> still the same, berry does not start

Then I added a print("test") statement to the beginning of web.be and berry started again. So with all the print() statements and without extra commented-out lines. Strange

**Comment by ChrisVonFrosta** [comment]
In general, I would appreciate a suggestion for a meaningful structure. To be honest, I don't even know the difference between driver and module. As mentioned many times, to add buttons I took the example from the Berry cookbook. There, only the driver class is created and an instance is created from it and registered as a driver. Is this then called cyclically? The driver class has no init function. i see no call of the class method that place the webbuttons. What I notice is that webserver.content_send("<p></p> test") permanently reaches an output in the web UI when set drew by the driver class in web.be. When called in module.be, the output disappears after 1 second. The former is placed in  buttons, the latter in sensor area.
I probably need more insight and overview, but I am also jumping around on 3 pages with Berry examples (cookbook, 20 minutes, wiki)...

**Comment by s-hadinger** [comment]
Can you share the code that causes the crash? In PM if you don't want to make it public

**Reply by ChrisVonFrosta** [reply]
i see no option to send PM over github?!

**Reply by barbudor** [reply]
You could use Discord

**Reply by ChrisVonFrosta** [reply]
that is possible. but i have to look for this failure script version.
i was in vacation last week and the night before, i did a lot of updates and workarounds. Since that i didn't got the error aggain.
